04.ls/ls.rb

```diff
@@ -1,19 +1,95 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'ffi-xattr'
+require 'etc'
+
 COLUMN_COUNT = 3
 SPACE = 1
+PERMISSION_TYPE = ['--x', '-w-', '-wx', 'r--', 'r-x', 'rw-', 'rwx'].freeze
 
 def main
   options = {}
   opt = OptionParser.new
-  opt.on('-a') { |v| v }
+  opt.on('-r')
+  opt.on('-a')
+  opt.on('-l')
   opt.parse!(ARGV, into: options)
 
-  path =  "#{Dir.getwd}/"
+  path = "#{Dir.getwd}/"
   entries = Dir.entries(path).sort
-  filtered_entries = options.key?(:a) ? entries : entries.reject { |entry| entry.start_with?('.') }
-  entries_with_suffix = append_suffix_by_file_type(filtered_entries, path)
+
+  sorted_entries = options.key?(:r) ? entries.reverse : entries
+  filtered_entries = options.key?(:a) ? sorted_entries : sorted_entries.reject { |entry| entry.start_with?('.') }
+  options.key?(:l) ? show_long_format(filtered_entries, path) : show_short_format(filtered_entries, path)
+end
+
+def show_long_format(entries, path)
+  long_format_entries = []
+  total_blocks = 0
+  entries.each do |entry|
+    lstat = File.lstat(entry)
+    total_blocks += lstat.blocks
+    long_format_entries << load_status_each_entry(entry, lstat, path)
+  end
+  print_long_data(long_format_entries, total_blocks)
+end
+
+def load_status_each_entry(entry, lstat, path)
+  xattr = Xattr::Lib.list(path + entry, @no_follow = true)
+  lstat_array = []
+  permission = read_permission(lstat)
+  lstat_array << (xattr.empty? ? "#{permission} " : "#{permission}@")
+  lstat_array << lstat.nlink.to_s.rjust(3)
+  lstat_array << Etc.getpwuid(lstat.uid).name
+  lstat_array << Etc.getgrgid(lstat.gid).name.rjust(5)
+  lstat_array << lstat.size.to_s.rjust(5)
+  lstat_array << lstat.ctime.strftime('%m %d %H:%M')
+  file_name = append_suffix_by_filetype(entry, path)
+  file_name_and_link_to = lstat.ftype == 'link' ? "#{file_name} -> #{File.readlink(entry)}" : file_name
+  lstat_array << file_name_and_link_to
+end
+
+def read_permission(lstat)
+  filetype_string =
+    case ftype = lstat.ftype
+    when 'fifo'
+      'p'
+    when 'file'
+      '-'
+    else
+      ftype.slice(0)
+    end
+  # File::lstat#modeの結果から権限に関わる部分を切り取る
+  for_permission_check = lstat.mode.to_s(8).rjust(6, '0')[3..6]
+  permission_string =
```

Comment:
```suggestion
  permission =
```

これもpermissionだけでよさそう